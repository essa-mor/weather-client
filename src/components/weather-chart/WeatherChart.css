.bottom_label{
	fill: #bababa;
	font-family: arial, sans-serif;
	font-size: 12px;
}

.bold{
	font-weight: bold;
}