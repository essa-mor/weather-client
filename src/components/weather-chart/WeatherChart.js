import React from 'react';

function WeatherChart({dt}) {
	return <div>Chart {dt}</div>;
}

export default WeatherChart;