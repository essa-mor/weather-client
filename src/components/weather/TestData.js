export const data = [
	{ dt: 1535274000, day: '2018-08-23 12:00:00', maxTemperature: 29.39, minTemperature: 27.66, img: '01d' },
	{ dt: 1535275000, day: '2018-08-24 12:00:01', maxTemperature: 30.39, minTemperature: 27.14, img: '01d' },
	{ dt: 1535276000, day: '2018-08-25 12:00:02', maxTemperature: 31.39, minTemperature: 26.95, img: '01d' },
	{ dt: 1535277000, day: '2018-08-26 12:00:03', maxTemperature: 32.39, minTemperature: 27.01, img: '02d' },
	{ dt: 1535278000, day: '2018-08-27 12:00:04', maxTemperature: 33.69, minTemperature: 26.9, img: '03d' }];