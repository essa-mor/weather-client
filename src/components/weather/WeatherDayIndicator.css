.weather_day_indicator{
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: space-around;
	height: 90px;
	width: 73px;
	cursor: pointer;
}

.weather_day_indicator_selected{
	background-color: #fcfcfc;
    border: 1px solid #e9e9e9;
    border-radius: 1px;
}

.weather_day{
	margin-top: 5px;
	font-family: arial, sans-serif;
	font-size: 13px;
	color: #bababa ;
}

.weather_temperatures{
	margin-bottom: 5px;
}
.weather_max_temperature{
	margin-right: 5px;
	font-family: arial, sans-serif;
	font-size: 13px;
	color: #878787 ;
}

.weather_min_temperature{
	font-family: arial, sans-serif;
	font-size: 13px;
	color: #bababa ;
}