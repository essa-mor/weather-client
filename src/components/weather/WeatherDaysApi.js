import mem from 'mem';

export function customFetch(url) {
	return fetch(url).then(response => response.json());
}

export function getMemCustomFetch(func, maxAge) {
	return mem(func, { maxAge });
}
const maxAge = 10 * 60 * 1000;
const memCustomFetch = getMemCustomFetch(customFetch, maxAge);

export function getWeatherDays() {
	return memCustomFetch('https://api.openweathermap.org/data/2.5/forecast/?q=Haifa,il&units=metric&APPID=8c6cf469699d615e37361375f85fb917')
		.then(data => getDailyWeather(data));
}

export function getWeatherDay(dt) {
	return memCustomFetch('https://api.openweathermap.org/data/2.5/forecast/?q=Haifa,il&units=metric&APPID=8c6cf469699d615e37361375f85fb917')
		.then(data => extractDayWeather(data, dt));
}

export function extractDayWeather(data, dt) {
	const result = [];
	const { list } = data;
	const firstIndex = list.findIndex(r => r.dt >= dt);
	for (let i = firstIndex; i < list.length && result.length < 8; i++) {
		const { dt_txt, dt } = list[i];
		const { temp } = list[i].main;
		result.push({ dt, dt_txt, temp });
	}
	return result;
}

export function getDailyWeather(data) {
	const result = [];
	const { list } = data;
	let currDtTxt = null;
	for (let i = 0; i < list.length; i++) {
		const { dt_txt, dt } = list[i];
		const { temp_max, temp_min } = list[i].main;
		const { icon } = list[i].weather[0];
		if (currDtTxt !== dt_txt.split(' ')[0]) {
			result.push({ dt, day: dt_txt, maxTemperature: temp_max, minTemperature: temp_min, img: icon });
			currDtTxt = dt_txt.split(' ')[0];
		}
	}
	return result;
}