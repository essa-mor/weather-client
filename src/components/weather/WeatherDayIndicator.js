import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './WeatherDayIndicator.css';

function WeatherDayIndicator({ isSelected, day, img, maxTemperature, minTemperature }) {
	const cls = classNames('weather_day_indicator', { 'weather_day_indicator_selected': isSelected });
	return (<div className={cls}>
		<span className="weather_day">{day}</span>
		<img src={img} alt="" />
		<div className="weather_temperatures">
			<span className="weather_max_temperature">{maxTemperature}°</span>
			<span className="weather_min_temperature">{minTemperature}°</span>
		</div>
	</div>);
}

WeatherDayIndicator.propTypes = {
	isSelected: PropTypes.bool,
	day: PropTypes.string.isRequired,
	img: PropTypes.string.isRequired,
	maxTemperature: PropTypes.number.isRequired,
	minTemperature: PropTypes.number.isRequired
};

export default WeatherDayIndicator;