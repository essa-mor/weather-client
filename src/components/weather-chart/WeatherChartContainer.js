import React from 'react';
import WeatherChart from './WeatherChart';
import { getWeatherDay } from '../weather/WeatherDaysApi';

class WeatherChartContainer extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = { data: [], dt: null };
	}

	fetchData(){
		const { match: { params: { dt } } } = this.props;
		debugger;
		if(dt !== this.state.dt && !isNaN(dt)){
			getWeatherDay(dt).then(data => this.setState({ data, dt }));
		}
	}

	componentDidMount() {
		this.fetchData();
	}

	componentDidUpdate() {
		this.fetchData();
	}

	render() {
		const { data } = this.state;
		return <WeatherChart data={data} />;
	}
}

export default WeatherChartContainer;