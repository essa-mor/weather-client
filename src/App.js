import React, { Component } from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import './App.css';
import WeatherDaysForecastContainer from './components/weather/WeatherDaysForecastContainer';
import WeatherChartContainer from './components/weather-chart/WeatherChartContainer';

class App extends Component {
	render() {
		return (
			<Router>
				<React.Fragment>
					<Route path="/:app/:dt" component={WeatherChartContainer} /> 
					<WeatherDaysForecastContainer/>
				</React.Fragment>
			</Router>
		);
	}
}

export default App;
