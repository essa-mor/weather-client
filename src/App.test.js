import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import App from './App';
import { getWeatherDays, getWeatherDay } from './components/weather/WeatherDaysApi';
import { data as daysData } from './components/weather/TestData';
import { data as dayData } from './components/weather-chart/TestData';

jest.mock('./components/weather/WeatherDaysApi');

describe('App test', () => {
	it('renders without crashing', () => {
		getWeatherDays.mockImplementation(() => Promise.resolve([]));
		const div = document.createElement('div');
		ReactDOM.render(<App />, div);
		ReactDOM.unmountComponentAtNode(div);
	});
	
	it('renders properly', async ()  => {
		getWeatherDays.mockImplementation(() => Promise.resolve(daysData));
		getWeatherDay.mockImplementation(() => Promise.resolve(dayData));
		const wrapper = shallow(<App />);
		await Promise.resolve();
		wrapper.update();
		expect(toJson(wrapper)).toMatchSnapshot();
	});	
});