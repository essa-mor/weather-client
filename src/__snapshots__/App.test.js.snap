// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test renders properly 1`] = `
<BrowserRouter>
  <Route
    component={[Function]}
    path="/:app/:dt"
  />
  <withRouter(WeatherDaysForecastContainer) />
</BrowserRouter>
`;
